src/auth/auth.service.ts

```diff
@@ -0,0 +1,156 @@
+import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
+import { JwtService } from '@nestjs/jwt';
+import * as bcrypt from 'bcrypt';
+import { UserService } from '../users/user.service';
+import { UserDto } from '../users/user.dto';
+import { User } from '../users/user.entity';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private userService: UserService,
+    private jwtService: JwtService
+  ) {}
+
+  async signUp(userDto: UserDto): Promise<User> {
+    await this.userService.createUser(userDto);
+    return await this.signIn(userDto);
+  }
+
+  async signIn(userDto: UserDto): Promise<User> {
+    const {email} = userDto;
+    const user: User = await this.userService.findUser({email});
+    const payload = { email: user.email, id: user.id };
+    user.token = this.jwtService.sign(payload);
+    return await this.userService.updateUser(user);
+  }
+
+  async signOut(token: string): Promise<User> {
+    const tokenData: any = this.jwtService.decode(token);
+    const {id, email} = tokenData;
+    const user: User = await this.userService.findUser({id, email});
+    user.token = null;
+    return await this.userService.updateUser(user);
+  }
+
+  validateToken(token: string): boolean {
+    try {
+      this.jwtService.verify(token);
+      return true;
+    } catch (err) {
+      throw new HttpException({
+        status: HttpStatus.FORBIDDEN,
+        errors: [err.message],
+      }, HttpStatus.FORBIDDEN);
+    }
+  }
+
+  async validateSignUpFields(userDto: UserDto): Promise<string[] | null> {
+    const errors = [];
+    const {username, email, password} = userDto;
+    const emailError: string = this.checkEmail(email);
+    const passwordError: string = this.checkPassword(password);
+    const usernameError: string = this.checkUsername(username);
+    const userError: string = await this.checkUserExistence(userDto);
+
+    if (emailError) {
+      errors.push(emailError);
+    }
+    if (passwordError) {
+      errors.push(passwordError);
+    }
+    if (usernameError) {
+      errors.push(usernameError);
+    }
+    if (userError) {
+      errors.push(userError);
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async validateSignInFields(userDto: UserDto): Promise<string[] | null> {
+    const errors = [];
+    const {email, password} = userDto;
+    const emailError: string = this.checkEmail(email);
+    const passwordError: string = this.checkPassword(password);
+    const userError: string = await this.checkUser(userDto);
+
+    if (emailError) {
+      errors.push(emailError);
+    }
+    if (passwordError) {
+      errors.push(passwordError);
+    }
+    if (userError && !errors.length) {
+      errors.push(userError);
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async validateSignOutFields(token: string): Promise<string[] | null> {
+    const errors = [];
+    const tokenData: any = this.jwtService.decode(token);
```

Comment:
it's not necessary to describe types in a local variable inside the function, especially if its _any_

Reply:
@haltaction here is errors without type:
![21-11-2020 17-46-52](https://user-images.githubusercontent.com/8670953/99881301-9e852180-2c21-11eb-9bae-7f98f8d6caed.png)

Reply:
It's probably the TSLint rule, usually, IDE allows to see the rule and change TSLint config. 
By the way, what is the full error message?

Reply:
@haltaction 
- `property 'email' does not exist on type 'string' | {[key: string]: any}`
- `property 'id' does not exist on type 'string' | {[key: string]: any}`